core/templates/domain/question/QuestionObjectFactory.ts

```diff
@@ -17,147 +17,150 @@
-    /* eslint-disable dot-notation */
-    Question['createDefaultQuestion'] = function(skillIds) {
-    /* eslint-enable dot-notation */
-      return new Question(
-        null, StateObjectFactory.createDefaultState(null),
-        DEFAULT_LANGUAGE_CODE, 1, skillIds);
-    };
-
-    Question.prototype.getValidationErrorMessage = function() {
-      var interaction = this._stateData.interaction;
-      if (interaction.id === null) {
-        return 'An interaction must be specified';
-      }
-      if (interaction.hints.length === 0) {
-        return 'At least 1 hint should be specfied';
+// require('domain/state/StateObjectFactory.ts');
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+import { StateObjectFactory, State } from 'domain/state/StateObjectFactory';
+
+export class Question {
+  id;
+  stateData;
+  languageCode;
+  version;
+  linkedSkillIds;
+  constructor(
+      id: number, stateData: State, languageCode: string, version: number,
+      linkedSkillIds: string[]) {
+    this.id = id;
+    this.stateData = stateData;
+    this.languageCode = languageCode;
+    this.version = version;
+    this.linkedSkillIds = linkedSkillIds;
+  }
+  getId(): number {
+    return this.id;
+  }
+  getStateData(): State {
+    return this.stateData;
+  }
+  getLanguageCode(): string {
+    return this.languageCode;
+  }
+  getVersion(): number {
+    return this.version;
+  }
+  getLinkedSkillIds(): string[] {
+    return this.linkedSkillIds;
+  }
+  setId(id: number): void {
+    this.id = id;
+  }
+  setStateData(stateData: State): void {
+    this.stateData = stateData;
+  }
+  setLanguageCode(languageCode: string): void {
+    this.languageCode = languageCode;
+  }
+  setVersion(version: number): void {
+    this.version = version;
+  }
+  setLinkedSkillIds(linkedSkillIds: string[]): void {
+    this.linkedSkillIds = linkedSkillIds;
+  }
+  getValidationErrorMessage(): string {
+    var interaction = this.stateData.interaction;
+    if (interaction.id === null) {
+      return 'An interaction must be specified';
+    }
+    if (interaction.hints.length === 0) {
+      return 'At least 1 hint should be specfied';
+    }
+    if (
+      !interaction.solution) {
+      return 'A solution must be specified';
+    }
+    var answerGroups = this.stateData.interaction.answerGroups;
+    var atLeastOneAnswerCorrect = false;
+    for (var i = 0; i < answerGroups.length; i++) {
+      if (answerGroups[i].outcome.labelledAsCorrect) {
+        atLeastOneAnswerCorrect = true;
+        continue;
       }
-      if (
-        !interaction.solution &&
-        INTERACTION_SPECS[interaction.id].can_have_solution) {
-        return 'A solution must be specified';
+    }
+    if (!atLeastOneAnswerCorrect) {
+      return 'At least one answer should be marked correct';
+    }
+    return null;
+  }
+
+  getUnaddressedMisconceptionNames(
+      misconceptionsBySkill) {
```

Comment:
Missing arg type. Ditto here and everywhere.

Reply:
Done, but StateObjectFactory and QuestionObjectFactory don't have predefined interfaces so I'm returning 'Object' currently (for toBackendDict). Should I implement Interfaces for these?